DEVELOPING.md

```diff
@@ -1,75 +1,236 @@
-
-Test that the build was successful by using an example model:
+The Toolbox distributables will be output in the `toolbox/org.lamport.tla.toolbox.product.product/target/products` directory.
+
+Developing with the Eclipse IDE
+-------------------------------
+
+You are free to use your own editor, but this project is largely developed in the [Eclipse](https://www.eclipse.org/) integrated development environment (IDE).
+
+To develop the TLA⁺ CLI tools in Eclipse, use Eclipse to open the `tlatools/org.lamport.tlatools` project.
+This Eclipse project configuration lives in the `tlatools/org.lamport.tlatools/.classpath` file.
+This file is also consumed by the Toolbox IDE build, so some local modifications are required before Eclipse can fully build & test the project.
+Copy & paste the following into a file called `classpath.patch` within your repo:
+```diff
+diff --git a/tlatools/org.lamport.tlatools/.classpath b/tlatools/org.lamport.tlatools/.classpath
+index c4c89e181..2caf43a97 100644
+--- a/tlatools/org.lamport.tlatools/.classpath
++++ b/tlatools/org.lamport.tlatools/.classpath
+@@ -26,11 +26,15 @@
+ 	<classpathentry kind="lib" path="lib/jline/jline-builtins-3.25.0.jar"/>
+ 	<classpathentry kind="lib" path="lib/jline/jline-reader-3.25.0.jar"/>
+ 	<classpathentry kind="lib" path="lib/jline/jline-terminal-3.25.0.jar"/>
+-	<classpathentry kind="lib" path="lib/jpf.jar"/>
+ 	<classpathentry kind="lib" path="lib/jmh/jmh-core-1.21.jar"/>
+ 	<classpathentry kind="lib" path="lib/jmh/commons-math3-3.2.jar"/>
+ 	<classpathentry kind="lib" path="lib/gson/gson-2.8.6.jar"/>
+ 	<classpathentry kind="lib" path="lib/lsp/org.eclipse.lsp4j.debug_0.21.1.v20230829-0012.jar"/>
+ 	<classpathentry kind="lib" path="lib/lsp/org.eclipse.lsp4j.jsonrpc_0.21.1.v20230829-0012.jar"/>
++	<classpathentry kind="lib" path="lib/junit-4.12.jar"/>
++	<classpathentry kind="lib" path="lib/jpf.jar"/>
++	<classpathentry kind="lib" path="lib/easymock-3.3.1.jar"/>
++	<classpathentry kind="lib" path="lib/aspectjtools-1.9.2.jar"/>
++	<classpathentry kind="lib" path="lib/hamcrest-core-1.3.jar"/>
+ 	<classpathentry kind="output" path="class"/>
+ </classpath>
+```
+Then, run `git apply classpath.patch`.
+You can right-click the root directory in Eclipse's Project Explorer and click Refresh to get Eclipse to pick up the `.classpath` changes.
+You will need to avoid including these changes in whatever commits you produce.
+One method is to run `git checkout .classpath` to revert the changes, create your commits, then again run `git apply classpath.patch`.
+
+When running unit tests in Eclipse, many of the tests depend on TLA⁺ module and model files in the repository.
+These tests need to know where to look for these files.
+For this reason, you will often need to modify the test run configurations by adding `-Dbasedir=${workspace_loc:tlatools}` to the VM arguments (not the program arguments) in the Arguments tab of the run configurations menu.
```

Comment:
`-Dbasedir...` was not always necessary.  It creeped with some recent-ish Eclipse IDE upgrade.  It would be good to eventually get rid of it.

Reply:
Bisecting identified ad454838ad3eaa296aad212e5e1d480b450ead37 as the starting commit after which `-Dbasedir=...` became necessary.

Reply:
@Calvin-L Any idea how to get the best in both/all worlds?

Reply:
Use the new [Path.resolve API](https://docs.oracle.com/javase/8/docs/api/java/nio/file/Path.html#resolve-java.lang.String-) instead of concatenating strings ourselves?

Reply:
Are you sure that that will work in Eclipse, ant, maven, IntelliJ, and elsewhere?

Reply:
How about this patch instead?

```diff
diff --git a/tlatools/org.lamport.tlatools/test/tlc2/tool/CommonTestCase.java b/tlatools/org.lamport.tlatools/test/tlc2/tool/CommonTestCase.java
index f24b805fb..452581ee9 100644
--- a/tlatools/org.lamport.tlatools/test/tlc2/tool/CommonTestCase.java
+++ b/tlatools/org.lamport.tlatools/test/tlc2/tool/CommonTestCase.java
@@ -54,7 +54,7 @@ import util.IsolatedTestCaseRunner;
 @RunWith(IsolatedTestCaseRunner.class)
 public abstract class CommonTestCase {
 
-       protected static final String BASE_DIR = System.getProperty("basedir", "");
+       protected static final String BASE_DIR = System.getProperty("basedir", System.getProperty("user.dir", "."));
        protected static final String TEST_MODEL = "test-model" + File.separator;
        public static final String BASE_PATH = System.getProperty("basepath", BASE_DIR + File.separator + TEST_MODEL);
 
```

(Generally it is safe to add extra `/` to a path; `/tmp/x` is the same as `/tmp///////x`, at least on Unix-based systems. The one exception is adding `/` to an empty string; `"" + "/" + "x"` is not the same as `"" + "x"`, since the first is an absolute path and the second is relative to the current working directory.)

Reply:
[`user.dir` is the magic system property for the current working directory](https://docs.oracle.com/javase/tutorial/essential/environment/sysprop.html)

Reply:
@Calvin-L Looks good.  Tests work without manually setting `basedir` with your patch.

Reply:
> Use the new [Path.resolve API](https://docs.oracle.com/javase/8/docs/api/java/nio/file/Path.html#resolve-java.lang.String-) instead of concatenating strings ourselves?

[Yes](https://github.com/tlaplus/tlaplus/blob/4613109641676389d97b8df209d6cf4d90d31c1c/tlatools/org.lamport.tlatools/src/tlc2/util/BufferedRandomAccessFile.java#L184). :) But independently of that, we have to decide what `basedir` even means, and how the tests should behave when it is not provided.

The Most Correct Fix might actually be even more invasive: make `test-model` a resource directory so it is available on the classpath during test execution. Do away entirely with file APIs and use Java's resource resolution mechanism instead. That would make the tests completely independent of the directory in which they're run. That is the most canonical way for Java tests to depend on data files, but it may require some careful tweaks to achieve.

Reply:
May I suggest to push @Calvin-L patch from above and remove the paragraph from DEVELOPING.md in this PR.

More substantial work is likely related to https://github.com/tlaplus/tlaplus/issues/719.

Reply:
https://github.com/tlaplus/tlaplus/pull/975

Reply:
With #975 merged, this paragraph can be removed from DEVELOPING.md.